### ISSUE TITLE
diamond is not py3k ready

### ISSUE THREAD

**Issue by sandrotosi** [python-diamond/Diamond#396]
python3 setup.py build  works fine but then running the result:

```
$ PYTHONPATH=build/lib/ python3 build/scripts-3.5/diamond
  File "build/scripts-3.5/diamond", line 112
    print "Diamond version %s" % (get_diamond_version())
                             ^
SyntaxError: invalid syntax
```

this is likely just the tip of the iceberg, and there is more work to do to make diamond really py3k ready

**Comment by mattrobenolt** [comment]
I can say there's almost 0 chance this will happen and almost 0 value for the project. iirc, Diamond also supports down to 2.5, so I think it's almost impossible to support from 2.5 all the way through 3.5 since there are some fundamental syntax changes.

I could be wrong here, but I just don't see the value in Diamond needing to run in python3 either. It's not a library, so it should be self contained and shouldn't need to interop with other code.

**Comment by sandrotosi** [comment]
python 2.5/2.6 are dead, so keeping their compatibility (and doing so preventing "progress") seems the wrong way for a python project.

other than then usual boilerplate of the advantage of py3k,  it is anyway the future of the language and a huge amount of work has been done to allow developers to write code that could be executed -without any change- both in 2.7 and 3.3+

There might be also people that would like to only run python3 on their machine (Arch switched /usr/bin/python to python3 long time ago), and given it's possible to extend with additional collectors, those might be easier written with a py3k-friendly syntax and/or you could gather more collaborations not closing that door in the code.

I wont use expressions like "0 value"/ no value, as it is discouraging and shortsighted.

thanks anyway for your attention

**Comment by jaingaurav** [comment]
@sandrotosi: So I think support python 3 is important due to environments like Arch (and I think the latest Ubuntu as well). However transitioning to python3 is a significant undertaking beyond the core codebase. We need to ensure that collectors and their dependent libraries are python3 compatible, at least for the collectors that are regularly supported.

In the end, the real argument is time. Many of the diamond developers are strapped for time and we currently have some core priorities that need to be addressed first before taking on such an undertaking. If you are willing to spend the time to work on the changes, I'd be more than happy to help you in whatever ability I can.

**Comment by sandrotosi** [comment]
@jaingaurav hehe i didnt mean "migrate it to py3k tomorrow" it's just something to keep in mind; during the last months/year I saw a huge number of project dropping support for 2.5/2.6 and 3.1/3.2 in favor of 2.7 and 3.3+ as they allow to have basically teh same code running on py2 and py3k.

so whenever some code is added or changed, you can think about this, instead of doing a mass transition, which can happen at a later point, once you are more used to write bilingual python.

Sadly, I dont have time either: I'm packaging diamond for debian, and the compat with py3k is one of my standard checks, hence this report

**Comment by jaingaurav** [comment]
@sandrotosi: Great stuff.

Btw, do you regularly package diamond for debian? Could you see if there is a easy fix for Issue #397?

**Comment by sandrotosi** [comment]
it was not in debian, so I did the first packaging, and I will continue to upgrade it as new releases will be cut (btw, do you plan to release soon? the version on pypi is rather old ;) ) - I think https://github.com/python-diamond/Diamond/issues/397 uses the debian/ dir inside diamond, while I wrote it from scratch (taking some bits and pieces when I needed from your packaging), you can have a look at http://anonscm.debian.org/viewvc/python-apps/packages/diamond/trunk/debian/ (ie, i would let debian/rules do the installation of README.md and not setup.py)

**Comment by snordhausen** [comment]
Please keep in mind that Python 2.6 is still alive and will remain so until the year 2020 due to [Red Hat Enterprise Linux 6](https://access.redhat.com/support/policy/updates/errata/#Life_Cycle_Dates). So you may not want to drop support for that.

There's already a pretty nice [pull request](https://github.com/python-diamond/Diamond/pull/259) that only drops support for 2.5 and still gets py3k working (almost). I really like this approach because it provides a big advantage (=py3k) with hardly any disadvantage (=drop only 2.5). In my experience this works out pretty well, even though you miss out on some of the cool features of 2.7.

And while some external components may not be compatible with py3k, officially dropping support for Python 2.5 would be the first step towards that goal. Because trying to get code to run under both 2.5 and py3k is pretty futile.

**Comment by jaingaurav** [comment]
@snordhausen: Agreed, python 2.6 support is not going away anytime soon. I'd like to get that PR in, but I think it wasn't updated after some initial comments. Perhaps if someone could resurrect it and give it a bit of massaging we could integrate it.

**Comment by redref** [comment]
Tried to get it working in #419 . Hope this can help.

**Comment by RichardBronosky** [comment]
This is really annoying. I can't find a single demo of this project. Not a screenshot. The first thing I run into when I finally decide to install it just to learn about it... is this. An insistence on not supporting Python 3 so that they can support 2.5. The reason we keep quoting Gretzky is because people still don't get it. "I skate to where the puck is going to be, not to where it has been." http://www.macleans.ca/economy/business/why-business-people-wont-stop-using-that-gretzky-quote/

**Comment by mattrobenolt** [comment]
> This is really annoying. 

Pull requests accepted I'm sure.

**Comment by josegonzalez** [comment]
@mattrobenolt 💯 I'd definitely immediately look at anyone's pull request to add Python3 support, while at least supporting Python 2.7 (which is where I currently deploy on Ubuntu 14.04). One minor blocker is getting as many bug fixes merged into the existing collectors, and then moving them out to their own projects, which should ease the upgrade burden. We'll still need to figure out packaging - debs, rpms, etc. - but thats a separate issue entirely.

We can make a major release as soon as that is the case.

**Comment by eirnym** [comment]
Is there any work going to bring Python3 compatibility?

**Comment by josegonzalez** [comment]
@eirnym Not at the moment, but if you'd like to contribute a pull request, I would love to review/merge it.

**Comment by hugovk** [comment]
I think dropping Python 2.6 (and earlier) would be a good first step. 2.6 is EOL and no longer receiving security updates, and is little used.

Here's the pip installs for Diamond from PyPI for last month:

| python_version | percent | download_count |
| -------------- | ------: | -------------: |
| 2.7            |  98.43% |         21,433 |
| 2.6            |   1.41% |            306 |
| 3.6            |   0.11% |             24 |
| 3.5            |   0.03% |              6 |
| 3.4            |   0.03% |              6 |

Source: `pypinfo --start-date -42 --end-date -15 --percent --pip --markdown Diamond pyversion`

Please could we get https://github.com/python-diamond/Diamond/pull/682 ("Remove unsupported Python 2.4 and 2.5 code") merged to begin with?

**Comment by josegonzalez** [comment]
@hugovk done!

**Comment by hugovk** [comment]
Thanks! See PR #703 to drop Python 2.6, do some updates for 2.7+, and prep towards Python 3.

**Comment by josegonzalez** [comment]
Merged that! If you can continue working on Py3, that would be radical :)

**Comment by jantman** [comment]
Python 2.7 is now EOL as well. I'm not sure if Diamond is even still maintained, but it seems like if this project is going to continue than py3 support is pretty much a necessity...
